DOCSP-20503: Enum Codec updates

ccho-mongodb opened the pull request:
## Pull Request Info

### Issue JIRA link:
https://jira.mongodb.org/browse/DOCSP-20503

### Snooty build log:
https://workerpool-boxgs.mongodbstitch.com/pages/job.html?collName=queue&jobId=6220e7b89c4c7f6bfb096ff7

### Docs staging link (requires sign-in on MongoDB Corp SSO):
[New Default Codec Registry section](https://docs-mongodbcom-staging.corp.mongodb.com/java/docsworker-xlarge/DOCSP-20503-enum-codec/fundamentals/data-formats/codecs/#default-codec-registry)
[What's New section](https://docs-mongodbcom-staging.corp.mongodb.com/java/docsworker-xlarge/DOCSP-20503-enum-codec/whats-new/)
[Enum type support](https://docs-mongodbcom-staging.corp.mongodb.com/java/docsworker-xlarge/DOCSP-20503-enum-codec/fundamentals/data-formats/pojo-customization/#enum-type-support)

### Self-Review Checklist

- [x] Is this free of any warnings or errors in the RST?
- [x] Did you run a spell-check?
- [ ] Did you run a grammar-check?
- [x] Does it render on staging correctly?
- [x] Are all the links working?
- [x] Are the staging and workerpool job links in the PR description updated?

### If your page documents a concept, does it meet the following criteria?

- [ ] Target the [Jasmin persona](https://drive.google.com/file/d/14FbBOLCVxwSP6M9BK4Nz1Ir9tzxT8_02/view)
- [ ] Target the [Lucas persona](https://drive.google.com/file/d/1J2vqJxo7ldv7OP_obA9Q-avf0o_ju4Lk/view)

terakilobyte commented on source/fundamentals/data-formats/codecs.txt:
```diff
@@ -225,8 +225,46 @@ their behavior. See our
+the following API documentation sections:
+
+- `CodecProvider <{+api+}/apidocs/bson/org/bson/codecs/configuration/CodecProvider.html>`__
+- `DefaultCodecRegistry <{+api+}/mongodb-driver-core/com/mongodb/MongoClientSettings.html#getDefaultCodecRegistry()>`__
```
issue: This link results in a 404.
```suggestion
- `DefaultCodecRegistry <{+api+}/apidocs/mongodb-driver-core/com/mongodb/MongoClientSettings.html#getDefaultCodecRegistry()>`__
```

terakilobyte commented on source/fundamentals/data-formats/pojo-customization.txt:
```diff
@@ -648,42 +650,46 @@ see the following API Documentation:
-field:
-
-.. code-block:: java
+The ``PojoCodecProvder`` includes a codec to convert the ``enum`` property
```
issue: typo

```suggestion
The ``PojoCodecProvider`` includes a codec to convert the ``enum`` property
```

terakilobyte commented on source/fundamentals/data-formats/pojo-customization.txt:
```diff
@@ -648,42 +650,46 @@ see the following API Documentation:
+.. note::
 
-   public class Penguin {
+   When using driver versions 4.5 and later, you can include the ``EnumCodec``
+   without the ``PojoCodecProvider``.
```
Should this be shown in the code below?

ccho-mongodb replied:
No, this page is generally about using the PojoCodecProvider. The admonition informs of a different case (working without the PojoCodecProvider) that is tangentially related. Since this is the only content that relates to enums and PojoCodecProvider, I wasn't able to find any other suitable location.

terakilobyte commented on source/fundamentals/data-formats/pojo-customization.txt:
```diff
@@ -648,42 +650,46 @@ see the following API Documentation:
 
-   Penguin penguin = new Penguin("Robin", 51, ADELIE);
+.. literalinclude:: /includes/fundamentals/code-snippets/codecs/EnumFieldPojo.java
+   :language: java
+   :dedent:
+   :start-after: start createPenguin
+   :end-before: end createPenguin
 
-The resulting document should look something like this:
+The resulting document should contain the following data:
```
Should this be an io codeblock?

ccho-mongodb replied:
No, it doesn't show output. It's describing the document structure, so JSON would be the best option.

terakilobyte reviewed (approved):
LGTM with a few corrections.

One overall question. According to the style guide, we should [use "To learn more" instead of "For more information"](https://docs.mongodb.com/meta/style-guide/style/links-and-cross-references/#link-and-cross-reference-text). I find them to be synonymous and don't have a preference. If you agree then we should let the team know.

ccho-mongodb commented:
> One overall question. According to the style guide, we should [use "To learn more" instead of "For more information"](https://docs.mongodb.com/meta/style-guide/style/links-and-cross-references/#link-and-cross-reference-text). I find them to be synonymous and don't have a preference. If you agree then we should let the team know.

Sure, I was aware of this but wasn't thinking about it at the time. I think I should revisit this with Tony as it doesn't seem like the style guide is currently following the guidance itself, and both options can lead to awkward repetition. Not able to figure out what the reasoning was from the PR that included the change (from a correction in preposition to a phrasing change):

https://github.com/mongodb/docs-meta/commit/d400dd001e3c42a87bd5398897aa39e2f5c6f770#diff-fbf00b12af5feeca0cef41d7c8bb2c6a65745ce7de7dc293fe0abc7834b556bdR57

ccho-mongodb commented:
I am going to leave "For more information..." in this PR because of the following:
1) It's not an addition, it's just moved and one link was added
2) The established pattern on the page is content + optional example(s) + API links introduced with "For more information", so it would break expectations to change only one and not all of them.

jyemin reviewed (changes requested):
I haven't finished reviewing, but wanted to provide the comments that I have so far.

ccho-mongodb commented on source/fundamentals/data-formats/codecs.txt:
```diff
@@ -225,8 +225,46 @@ their behavior. See our
+
+   CodecRegistry newRegistry = CodecRegistries.fromRegistries(
+       MongoClientSettings.getDefaultCodecRegistry(),
+       new MyCodecRegistry());
```
I see. Would `CodecRegistries.fromCodecs(new myCodec())` would be more relatable?

jyemin replied:
yes

ccho-mongodb commented on source/fundamentals/data-formats/codecs.txt:
```diff
@@ -225,8 +225,46 @@ their behavior. See our
+.. _codecs-override:
+
+If you need to override the behavior of a ``Codec`` for a type defined in
+the default codec registry, ensure that you place your custom ``Codec``
```
Sorry, I misread the comment in the ticket.

ccho-mongodb commented:
Addressed your feedback in the [latest commit](https://github.com/mongodb/docs-java/pull/211/commits/13e58d4121df6a051760852e15c732e2dc8943ba)

jyemin commented on source/fundamentals/data-formats/pojo-customization.txt:
```diff
@@ -648,42 +650,46 @@ see the following API Documentation:
-field:
-
-.. code-block:: java
+The ``PojoCodecProvider`` includes a codec to convert the ``enum`` property
```
As of 4.5, this is no longer part of the PojoCodecProvider.  Rather, the PojoCodecProvider delegates to the CodecRegistry to find a Codec for any enum class.  

In practice, applications are unlikely to use enum classes outside of the context of POJOs or the upcoming support for Java records.

ccho-mongodb replied:
My mistake, I thought the new codec was included. Would a good way to summarize the "Enum Type Support" for 4.5 in this section (within the POJO customization page) be to mention that as of 4.5, you must provide an enum Codec from another provider?

jyemin replied:
No, the enum code is included in the default registry, but independently of the pojo codec.  The pojo codec used to support enums directly.  Now it delegates to the registry.  So if an application happened to register an enum codec in the registry (unlikely, but possible), it will be used by the pojo codec.

jyemin reviewed (changes requested):
Are you planning to include my suggest release notes:

"The default codec registries accessible from MongoClientSettings class and Bson interfaces now include a codec that will encode and decode all enumeration classes. If your application is creating a registry based on either of these default registries, and your application's new registry contains a codec provider for enumerated classes, you must ensure that the provider for the enumerated classes precedes the default provider"

?

ccho-mongodb commented:
> Are you planning to include my suggest release notes:
> 
> "The default codec registries accessible from MongoClientSettings class and Bson interfaces now include a codec that will encode and decode all enumeration classes. If your application is creating a registry based on either of these default registries, and your application's new registry contains a codec provider for enumerated classes, you must ensure that the provider for the enumerated classes precedes the default provider"
> 
> ?

I thought that I had communicated all the items clearly in my additions to "What's New". Is there anything missing or unclear there?

https://github.com/mongodb/docs-java/pull/211/files#diff-a428c7afd1dc66fddc6d7a94de108ea11392d5ab3516b074233d538b50e6863aR42-R47

jyemin commented:
> I thought that I had communicated all the items clearly in my additions to "What's New". Is there anything missing or unclear there?

My bad.  I missed it.

ccho-mongodb commented:
Addressed feedback by deferring to the default codec registry section from the POJO enum types support section. Let me know if this organization of info makes sense. @jyemin

jyemin reviewed (approved):
LGTM
